[16.0][ADD] l10n_es_verifactu_pos_oca

ljsalvatierra-factorlibre opened the pull request.

ljsalvatierra-factorlibre committed:
[ADD] l10n_es_verifactu_pos

ljsalvatierra-factorlibre committed:
[FIX] l10n_es_verifactu_pos: Company VAT for receipt QR URL

ljsalvatierra-factorlibre committed:
[REF] l10n_es_verifactu_pos: Rename to l10n_es_verifactu_pos_oca

aritzolea commented on l10n_es_verifactu_pos_oca/models/pos_order.py:
```diff
@@ -0,0 +1,674 @@
+
+    def write(self, vals):
+        """Override write to protect fields once sent to verifactu"""
+        for _order in self.filtered(
```
Es necesario iterar aquí si no se comprueba nada de self?
Tal vez valdría por
```
if self.filtered(lambda x: x.verifactu_enabled and x.aeat_state != "not_sent"):
     # comprobaciones en vals
```

aritzolea commented on l10n_es_verifactu_pos_oca/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+    "application": False,
+    "installable": True,
+    "depends": [
+        "point_of_sale",
```
Podrías quitar esto, ya que `l10n_es_pos` depende ya de `point_of_sale`

aritzolea commented:
Revisa los test, parece que está chocando con algo del módulo `test_l10n_es_pos_sii`. Tal vez alguna función con el mismo nombre?

ljsalvatierra-factorlibre committed:
[REF] l10n_es_verifactu_pos_oca: changes after CR

ljsalvatierra-factorlibre commented:
> Revisa los test, parece que está chocando con algo del módulo `test_l10n_es_pos_sii`. Tal vez alguna función con el mismo nombre?

Resuelto, muchas gracias!

ljsalvatierra-factorlibre committed:
[FIX] l10n_es_verifactu_oca: search document when creating response

ljsalvatierra-factorlibre committed:
[FIX] l10n_es_verifactu_oca: search document when creating response

anmarmo1 reviewed (approved):
A nivel funcional lo veo correcto, solo que no había configurado la agencia tributaria y no se ha quejado, y al crear una factura si. No se si afecta al envío, lo pongo por si acaso.

MMartinezDA commented:
Hemos detectado que el modulo genera un archivo con el estatus "Aceptado con errores" cuando se hace un reembolse desde el punto de venta, desde la página del aeat se ve que algunos campos están en positivo y otro en negativo, pero cuando todos son negativos si lo acepta correctamente. 
Si el reembolso se hace desde el backend el modulo no envía ni genera el QR para verifactu, seguramente porque de esta forma no se le asignan al ticket un secuencia de la factura simplificada tampoco.

almumu committed:
[FIX] l10n_es_verifactu_pos_oca: correct amounts on refund pos orders

[FIX] l10n_es_verifactu_pos_oca: correct amounts on refund pos orders
[IMP] l10n_es_verifactu_pos_oca: clean unnecessary fields already defined on verifactu.mixin

CrazyIFriT commented:
He instalado el pull realmente hacer una venta desde el POS va genial, pero al ejecutar manualmente si es verdad que sale un error.

  raise ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr))
ValueError: <class 'psycopg2.errors.NotNullViolation'>: "el valor null para la columna «entry_id» viola la restricción not null
DETAIL:  La fila que falla contiene (1108, null, 1286, 0, 1, 1, null, incorrect, 3000, {
    'IDFactura': {
        'IDEmisorFactura': 'CIF',
   ..., 2025-10-29 11:16:07.435028, 2025-10-29 11:16:07.435028).
" while evaluating
'model._cron_send_documents_to_verifactu()'

Me supongo que es la compatiblidad con el pos, he podido en modo test enviar facturas sin problemas, pero esto es al instalar el pull para testear el POS

IvanHRN commented:
Por mi parte solo he encontrado un problema, y es que ante un envío erróneo no existe la opción de volver e reenviar el ticket. Una posible mejora también sería añadir filtros para buscar los envíos erróneos. 

Por lo demás he probado distintas posiciones fiscales, rectificativas desde tpv, facturas ordinarias desde tpv y todo correcto. Solo me ha devuelto un error hacienda al hacer una venta de sujeto pasivo nacional pero no es por el desarrollo en sí sino que hacienda no permite la inversión de sujeto pasivo con tipo de factura simplificada (F2). Si se realiza una factura NO simplificada el envío es correcto.

Lo que comenta @MMartinezDA de las rectificativas no lo he conseguido reproducir. En mi caso me envía los datos en negativo correctamente. Quizá es algo que está solucionado ya (el comentario es de hace 3 semanas)

"TipoFactura": "R5",
        "DescripcionOperacion": "TEST",
        "Desglose": {
            "DetalleDesglose": [
                {
                    "Impuesto": "01",
                    "ClaveRegimen": "01",
                    "CalificacionOperacion": "S1",
                    "TipoImpositivo": "21.0",
                    "BaseImponibleOimporteNoSujeto": -3.79,
                    "CuotaRepercutida": -0.8
                }
            ]
        },
        "CuotaTotal": -0.8,
        "ImporteTotal": -4.59, 

Si que es verdad lo que comenta al hacer la rectificativa desde el backend. Entiendo que esto no se debería de permitir por lo que quizá sería conveniente ocultar el botón para evitar problemas.

Por último, lo que comenta @CrazyIFriT tampoco lo he conseguido reproducir.

almumu commented:
@IvanHRN el poder reenviar un pedido del pos a verifactu (subsanaciones), lo estamos preparando, pero está a la espera de ver cómo progresa este otro pr https://github.com/OCA/l10n-spain/pull/4429 ya que se basa en eso.

Lo otro que comentan de las devoluciones creadas desde fuera del tpv,es cierto, da error el envío a verifactu porque no tiene número de factura simplificada la devolución, y no sé hasta qué punto podemos hacer algo aquí, porque el problema no es el envío, sino que no se asigna la secuencia, y eso pasa tengas instalado verifactu o no lo tengas, quizá es un issue del módulo l10n_es_pos, que no contempla este caso de hacerlo desde fuera del tpv.

IvanHRN commented:
Perfecto.

En cuanto a lo que comentas de las devoluciones estoy de acuerdo, quizá es un problema a resolver en el l10_es_pos.

pedrobaeza commented on l10n_es_verifactu_oca/models/verifactu_invoice_entry.py:
```diff
@@ -338,20 +338,13 @@ def _create_response_lines(
-                    break
-            # Find the verifactu.invoice entry for this document
-            verifactu_invoice_entry = document.last_verifactu_invoice_entry_id
+            matching_entries = self.filtered(lambda r: r.document_name == invoice_num)
```
Esto y lo siguiente lo hemos refinado en #4520 para 14. Extraedlo por favor en un PR aparte y con los cambios correspondientes, y lo fusionamos antes de ir con este otro módulo.

ljsalvatierra-factorlibre replied:
Nos ponemos con ello, gracias!

acysos commented:
Hola,

Veo que se ha modificado el recibo que se imprime pero el que muestra en pantalla y manda por mail, aquí no se muestra el QR.

Saludos

CrazyIFriT commented:
> Hola,
> 
> Veo que se ha modificado el recibo que se imprime pero el que muestra en pantalla y manda por mail, aquí no se muestra el QR.
> 
> Saludos

Hola

Haciendo pruebas por mi parte, el ticket que imprime por pantalla si sale el QR y el email igual, lo que tuve que activar la secuencia de factura simplificada en los ajustes en la 16.

CrazyIFriT commented:
Lo que sí no aparece en en la parte de pedidos si se quiere volver a reimprimir el ticket no sale el qr. No sé si soy el único que le pasa

almumu committed:
[FIX] l10n_es_verifactu_pos_oca: fix altas por rechazo y subsanación (#89)

* [FIX] l10n_es_verifactu_pos_oca: fix altas por rechazo y subsanación

Se arregla el envío de pos_orders cuando han sido previamente rechazados.

almumu committed:
[IMP] l10n_es_verifactu_pos_oca: improove verifactu views in pos order (tree, form and search)

alvarosalinero commented:
> He instalado el pull realmente hacer una venta desde el POS va genial, pero al ejecutar manualmente si es verdad que sale un error.
> 
> raise ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)) ValueError: <class 'psycopg2.errors.NotNullViolation'>: "el valor null para la columna «entry_id» viola la restricción not null DETAIL: La fila que falla contiene (1108, null, 1286, 0, 1, 1, null, incorrect, 3000, { 'IDFactura': { 'IDEmisorFactura': 'CIF', ..., 2025-10-29 11:16:07.435028, 2025-10-29 11:16:07.435028). " while evaluating 'model._cron_send_documents_to_verifactu()'
> 
> Me supongo que es la compatiblidad con el pos, he podido en modo test enviar facturas sin problemas, pero esto es al instalar el pull para testear el POS


Hola,

me pasa exactamente lo mismo. ¿Ya está solucionado?

Gracias.

Un saludo,

Álvaro

CrazyIFriT commented:
> > He instalado el pull realmente hacer una venta desde el POS va genial, pero al ejecutar manualmente si es verdad que sale un error.
> > raise ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)) ValueError: <class 'psycopg2.errors.NotNullViolation'>: "el valor null para la columna «entry_id» viola la restricción not null DETAIL: La fila que falla contiene (1108, null, 1286, 0, 1, 1, null, incorrect, 3000, { 'IDFactura': { 'IDEmisorFactura': 'CIF', ..., 2025-10-29 11:16:07.435028, 2025-10-29 11:16:07.435028). " while evaluating 'model._cron_send_documents_to_verifactu()'
> > Me supongo que es la compatiblidad con el pos, he podido en modo test enviar facturas sin problemas, pero esto es al instalar el pull para testear el POS
> 
> Hola,
> 
> me pasa exactamente lo mismo. ¿Ya está solucionado?
> 
> Gracias.
> 
> Un saludo,
> 
> Álvaro

Creo que es algo de la instalación, hay gente que no tiene estos errores y no los puede reproducir, lo que hice yo fue reinstalar el verifactu y el verifactu pos.

Saludos

almumu committed:
[FIX] l10n_es_verifactu_pos_oca: rebase branch and fix error PosOrder._get_verifactu_hash_string() got an unexpected keyword argument 'cancel'

ozono reviewed (approved):
Probado. Válido para mergear y trabajar en la rama oficial.

HaraldPanten commented:
@ljsalvatierra-factorlibre @almumu En qué estado está el PR? Entiendo que ya lo habéis implementado en producción con alguno de vuestros clientes, correcto?

Podéis hacer rebase, por favor?

Gracias por la aportación.

almumu commented:
> @ljsalvatierra-factorlibre @almumu En qué estado está el PR? Entiendo que ya lo habéis implementado en producción con alguno de vuestros clientes, correcto?
> 
> Podéis hacer rebase, por favor?
> 
> Gracias por la aportación.

Buenos días @HaraldPanten , rebase hecho. En teoría este PR estaba revisado y listo para mergear. @pedrobaeza  sugirió extraer la parte donde se modifican ficheros del módulo l10n_es_verifactu_oca a un PR a parte (https://github.com/OCA/l10n-spain/pull/4378#discussion_r2482471388) , se hizo otro PR https://github.com/OCA/l10n-spain/pull/4659 donde además de mejorar esa parte, se incluía un cherry-pick de un fix que no se había aplicado en V16 de verifactu, pero no hubo más avances con eso.

ozono commented:
@pedrobaeza ya que estás activo con otros PRs de Veri*FACTU estos días, ¿sería un buen momento para mergear este? Está probado en producción permitiría darle soporte oficial y seguir mejorándolo en la rama principal. Gracias.

pedrobaeza commented:
@ljsalvatierra-factorlibre puedes por favor hacer squash de todos los commits en uno? Al final, ese histórico no está aportando valor en el estado actual.

ljsalvatierra-factorlibre committed:
[ADD] l10n_es_verifactu_pos_oca: VeriFactu support for Point of Sale

Add immediate submission of POS invoicing to the AEAT VeriFactu system.
The module integrates the Point of Sale with l10n_es_verifactu_oca:

* computes the VeriFactu fingerprint (hash chain) for POS orders
* renders the VeriFactu QR code and fiscal data on the customer receipt
* adds VeriFactu fields and views to pos.order (tree, form and search)
* includes tests covering the POS order flow

Co-authored-by: almumu <almumu@users.noreply.github.com>

almumu commented:
> @ljsalvatierra-factorlibre puedes por favor hacer squash de todos los commits en uno? Al final, ese histórico no está aportando valor en el estado actual.

squash hecho @pedrobaeza

pedrobaeza commented:
/ocabot merge patch

HaraldPanten reviewed (approved):
/ocabot merge nobump

HaraldPanten commented:
/ocabot merge nobump

HaraldPanten commented:
/ocabot merge nobump

pedrobaeza commented:
@sbidoul maybe you can create this project by hand, as we are trying a lot of times, and it's impossible. It seems the limit is even lower than 4, as I don't remember these days to merge that quantity.

sbidoul commented:
added

pedrobaeza commented:
/ocabot merge nobump
